core/lib/instrument_api.h

```diff
@@ -1806,6 +1806,19 @@ DR_API
 const char *
 dr_get_application_name(void);
 
+DR_API
+/**
+ * Sets the passed arguments to the command-line arguments of the
+ * application.
+ * @param argc The pointer where to place argc.
+ * @param argv The pinter where to place argv.
+ * @return true if successful.
+ * \note Currently, this function is only available on Unix with
+ * early injection.
+ */
+bool
+dr_get_app_args(OUT int **argc, OUT char ***argv);
```

Comment:
I'm not sure we resolved the discussion from PR #2885: https://github.com/DynamoRIO/dynamorio/pull/2885#discussion_r175331626

Reviving it here:

On Windows the command line is in RTL_USER_PROCESS_PARAMETERS.CommandLine as a UNICODE_STRING. It is not parsed into tokens and there is no count of tokens: so there's no `argc`, and the UNICODE_STRING is a data structure with a size and a pointer to content.  What is this function going to do on Windows?  Will the interface have to change?  Is `argv` going to point to the UNICODE_STRING struct, or to the array of UTF-16 characters its field points to?  Neither of those is anything like a `char **` array of *pointers* to token strings.  Should this at least have a type of `void *`?

Reply:
Didn't think much about it but how about: 1) we forget about argc 2) change the function's prototype to dr_get_app_args(OUT void **args, OUT size_t size) 3) for linux we pass the array of args and size being equal to argc, 4) for windows pass an array with one element, namely UNICODE_STRING, and set size to 1?

Reply:
OK, but I would tweak it for Windows since UNICODE_STRING is a native-library-level data structure and not exposed at the Windows API level: I would point the data at the UTF-16 string and set the size to the size of that string (in bytes, not chars (yes it's different since it's UTF-16): just read the struct field).

Reply:
Seems like a reasonable solution.  STGM.

Reply:
> I would point the data at the UTF-16 string and set the size to the size of that string (in bytes, not chars (yes it's different since it's UTF-16): just read the struct field).

But then we would have size being used for different purposes. On Unix size would denote the count of args, and on Windows it would denote the number of bytes. I don't think that is what one wants. Can't we rely on NULL terminations to get the length?  s/size/count -> dr_get_app_args(OUT void **args, OUT size_t *count)

Reply:
> On Unix size would denote the count of args, and on Windows it would denote the number of bytes.

Let's make UNIX be the number of bytes too, then.  It's like the "memory blob" discussed previously: `dr_get_app_args` returns a region of memory that contains the application command line, in the form of the start address and the size in bytes.  The caller has to interpret it on a per-platform basis (which is not ideal: DR normally tries to help the user write cross-platform code).

Maybe the name should be `dr_get_app_arg_loc` (or spell out `location`) to make it clearer, and to allow for adding a `dr_get_app_args` in the future which tokenizes WIndows.

Reply:
> Can't we rely on NULL terminations to get the length?

No, UNICODE_STRING strings are not always NULL-terminated.  Usually, but we have seen enough that aren't that it is not to be relied upon.

Reply:
Then how about the function returns an array of a new struct, something like dr_app_args_t, which consists of a string (char *) and the string's size in bytes (size_t). Maybe dr_get_app_args(dr_app_args_t *args, size_t *count).

Very easy to tokenise for Unix. Perhaps a bit challenging for Windows due to UTF-16 but the client is going to have to tokenise anyways if we just return the UTF-16 string. Might as well we bite the bullet and alleviate this burden for the user. 

Any idea how this UTF-16 string on windows structured please? Is it easy to parse?

Reply:
To clarify: so this is going to provide an array of argument token strings for both platforms, but it's trying to avoid making a copy, and that's why it's using a size and not null-terminating?  I think each token will be hard to work with if it's not null-terminated: regular C string routines won't work.  I would think we'd return an array of null-terminated strings, which would be copies on Windows.

Should the Windows copies be converted to UTF-8 so the type can just be "char *"?  The core already has code to convert back and forth.

The Windows string is what you would expect: the full command line just all run together with spaces.  We'd want to do what sthg like https://docs.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw does.  I think the challenging part is identifying unicode spaces and quotes without using library routines -- or finding ntdll library routines we can prove are safe to use.  Or maybe moving this to drx and trying to use that shell32 routine, but dynamically so we don't pull it in every time??  Or just implementing an ASCII-oriented version and documenting the limitations?

Reply:
Maybe it's fine to limit to ASCII spaces and quotes: DR is already doing that in some places and has our_isspace.  I don't know if we'll run into irregular spaces.  Yeah, I would say, let's just do that for now.

Reply:
So I am suggesting:
+ Allocate memory for copies converted to UTF-8 on Windows: so need some way to free
+ Only support parsing based on ASCII spaces and quotes

Reply:
Thanks. I understood what you mean. 

> To clarify: so this is going to provide an array of argument token strings for both platforms, but it's trying to avoid making a copy, and that's why it's using a size and not null-terminating?

Yes. In particular, I don't want to lose the tokens (on Unix) by just returning an opaque blob.

>  I think each token will be hard to work with if it's not null-terminated: regular C string routines won't work. I would think we'd return an array of null-terminated strings, which would be copies on Windows.

I propose the following: 
1) Apart from the pointer and the size, add a new boolean field to the dr_app_args_t struct, something like `is_utf_16`.
2) Create an additional API function, with a prototype  like dr_app_arg_as_string(IN dr_app_args_t  *app_arg, OUT char *buf_out, IN size_buf_size) This function takes an app_arg from the array which the user obtained from dr_get_app_args(), and sets buf_out with the corresponding null-terminated string. 

The function will do the copying and the  UTF-8 conversation (as you mentioned) when `is_utf_16` is true and notably will avoid issues regarding the freeing of the string buffer because this buffer is supplied by the user. For Unix, the function simply calls a strncpy.

The only downside with the approach is that dr_get_app_args() will also do the copying and conversion. on Windows, purely for the sake of parsing. But the upside is that the user gets to know the locations of args on an individual basis. 

> 
> Should the Windows copies be converted to UTF-8 so the type can just be "char *"? The core already has code to convert back and forth.
> 
> The Windows string is what you would expect: the full command line just all run together with spaces. We'd want to do what sthg like https://docs.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw does. I think the challenging part is identifying unicode spaces and quotes without using library routines -- or finding ntdll library routines we can prove are safe to use. Or maybe moving this to drx and trying to use that shell32 routine, but dynamically so we don't pull it in every time?? Or just implementing an ASCII-oriented version and documenting the limitations?

Reply:
> The only downside with the approach is that dr_get_app_args() will also do the copying and conversion on Windows, purely for the sake of parsing.

We should be able to parse to find the token boundaries without copying.

Reply:
OK so this seems good to me: one function which returns a list of boundaries and encoding of each token, and another which converts to utf-8.  Maybe the only thing to settle is how important the 2nd function is: the user could do that on their own today with dr_snprintf and the `%S` specifier.  Maybe writing some sample client code would help decide?  Maybe the 2nd function should return a char* so it can return a pointer to the original and not do a memcpy (documenting this of course), so it would return either a pointer to the original or to the passed-in buffer?  That might make user code simpler.

Reply:
> I think the challenging part is identifying unicode spaces and quotes without using library routines -- or finding ntdll library routines we can prove are safe to use

> We should be able to parse to find the token boundaries without copying.

I thought we needed to convert to utf-8 to help identify spaces? or did I misunderstand that part?

Reply:
If we want to pass to some library routine that takes utf-8: but if we limit to ascii we can identify in utf-16.

This is what it is sounding like to me, cross-platform code which as no memcpy on UNIX:

```C
    dr_app_arg_t arg_info[MAX_ARGS];
    size_t num = dr_get_app_args(&arg_info, sizeof(arg_info));
    for (int i = 0; i < num; ++i) {
        char arg_buf[MAX_ARG_SIZE];
        const char *arg_str = dr_app_arg_as_utf8(&arg_info[i], sizeof(arg_buf));
        /* now examine arg_str with regular strcmp, etc. functions */
    }
```

Reply:
Yep, we are on the same page.

Reply:
Done.